ur_robot_driver/src/hardware_interface.cpp

```diff
@@ -477,4 +575,95 @@ void URPositionHardwareInterface::handleRobotProgramState(bool program_running)
   }
   robot_program_running_ = program_running;
 }
+
+void URPositionHardwareInterface::initAsyncIO()
+{
+  for (size_t i = 0; i < 18; ++i)
+  {
+    standard_dig_out_bits_cmd_[i] = static_cast<double>(actual_dig_out_bits_[i]);
+  }
+
+  for (size_t i = 0; i < 2; ++i)
+  {
+    standard_analog_output_cmd_[i] = standard_analog_output_[i];
+  }
+}
+
+void URPositionHardwareInterface::checkAsyncIO()
+{
+  for (size_t i = 0; i < 18; ++i)
+  {
+    if (standard_dig_out_bits_cmd_[i] != static_cast<double>(actual_dig_out_bits_[i]))
+    {
+      if (i <= 7)
+      {
+        io_async_success_ =
+            ur_driver_->getRTDEWriter().sendStandardDigitalOutput(i, static_cast<bool>(standard_dig_out_bits_cmd_[i]));
+      }
+      else if (i <= 15)
+      {
+        io_async_success_ = ur_driver_->getRTDEWriter().sendConfigurableDigitalOutput(
+            static_cast<uint8_t>(i - 8), static_cast<bool>(standard_dig_out_bits_cmd_[i]));
+      }
+      else
+      {
+        io_async_success_ = ur_driver_->getRTDEWriter().sendToolDigitalOutput(
+            static_cast<uint8_t>(i - 16), static_cast<bool>(standard_dig_out_bits_cmd_[i]));
+      }
+
+      if (io_async_success_ != 2.0)
+        return;
+    }
+  }
+
+  for (size_t i = 0; i < 2; ++i)
+  {
+    if (standard_analog_output_cmd_[i] != standard_analog_output_[i])
+    {
+      io_async_success_ = ur_driver_->getRTDEWriter().sendStandardAnalogOutput(i, standard_analog_output_cmd_[i]);
+
+      if (io_async_success_ != 2.0)
```

Comment:
Replace with enum

Reply:
This is OK. There is no enum for this. Let's keep this as it is for the time being. Otherwise, will get tricky with dependencies between packages.